File: Beecrowd_Parte2/Fusões_2380.c
#include <stdio.h>
#include <stdlib.h>

// Definindo a estrutura Union-Find
typedef struct {
    int *parent;
    int *rank;
} UnionFind;

// Função para inicializar a estrutura Union-Find
UnionFind* createUnionFind(int size) {
    UnionFind* uf = (UnionFind*) malloc(sizeof(UnionFind));
    uf->parent = (int*) malloc((size + 1) * sizeof(int));
    uf->rank = (int*) malloc((size + 1) * sizeof(int));

    for (int i = 1; i <= size; i++) {
        uf->parent[i] = i;
        uf->rank[i] = 0;
    }

    return uf;
}

// Função para encontrar o representante (raiz) de um conjunto
int find(UnionFind* uf, int u) {
    if (uf->parent[u] != u) {
        uf->parent[u] = find(uf, uf->parent[u]); // Path compression
    }
    return uf->parent[u];
}

// Função para unir dois conjuntos
void unionSets(UnionFind* uf, int u, int v) {
    int rootU = find(uf, u);
    int rootV = find(uf, v);

    if (rootU != rootV) {
        // Union by rank
        if (uf->rank[rootU] > uf->rank[rootV]) {
            uf->parent[rootV] = rootU;
        } else if (uf->rank[rootU] < uf->rank[rootV]) {
            uf->parent[rootU] = rootV;
        } else {
            uf->parent[rootV] = rootU;
            uf->rank[rootU]++;
        }
    }
}

// Função principal
int main() {
    int N, K;
    scanf("%d %d", &N, &K);
    getchar(); // Consumir o newline após os inteiros

    UnionFind* uf = createUnionFind(N);

    for (int i = 0; i < K; i++) {
        char type;
        int A, B;
        scanf(" %c %d %d", &type, &A, &B);

        if (type == 'F') {
            unionSets(uf, A, B);
        } else if (type == 'C') {
            if (find(uf, A) == find(uf, B)) {
                printf("S\n");
            } else {
                printf("N\n");
            }
        }
    }

    // Liberar memória
    free(uf->parent);
    free(uf->rank);
    free(uf);

    return 0;
}

File: Beecrowd_Parte2/Fusões_2380.java
import java.util.*;

public class Main {

    static class UnionFind {
        int[] parent;
        int[] rank;

        public UnionFind(int size) {
            parent = new int[size + 1];
            rank = new int[size + 1];
            for (int i = 1; i <= size; i++) {
                parent[i] = i;
                rank[i] = 0;
            }
        }

        public int find(int u) {
            if (parent[u] != u) {
                parent[u] = find(parent[u]); // Path compression
            }
            return parent[u];
        }

        public void union(int u, int v) {
            int rootU = find(u);
            int rootV = find(v);
            
            if (rootU != rootV) {
                // Union by rank
                if (rank[rootU] > rank[rootV]) {
                    parent[rootV] = rootU;
                } else if (rank[rootU] < rank[rootV]) {
                    parent[rootU] = rootV;
                } else {
                    parent[rootV] = rootU;
                    rank[rootU]++;
                }
            }
        }
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        int N = scanner.nextInt();
        int K = scanner.nextInt();
        scanner.nextLine(); // Consumir o newline após os inteiros

        UnionFind uf = new UnionFind(N);

        for (int i = 0; i < K; i++) {
            String line = scanner.nextLine();
            String[] parts = line.split(" ");

            char type = parts[0].charAt(0);
            int A = Integer.parseInt(parts[1]);
            int B = Integer.parseInt(parts[2]);

            if (type == 'F') {
                uf.union(A, B);
            } else if (type == 'C') {
                if (uf.find(A) == uf.find(B)) {
                    System.out.println("S");
                } else {
                    System.out.println("N");
                }
            }
        }

        scanner.close();
    }
}
